### PR TITLE
Adding octo config 

### PR DESCRIPTION
**Pull request by prayagsingh**
close #750

### CODE REVIEW

File: jvb/Dockerfile [hunk 1]
```diff
@@ -8,4 +8,13 @@ RUN \
...
 
+# https://github.com/jitsi/docker-jitsi-meet/issues/207
+RUN \
+   apt-dpkg-wrap apt-get update && \
+   apt-dpkg-wrap apt-get install -y wget && \
+  	apt-cleanup && \
+   wget -O /ssl.deb http://security.debian.org/debian-security/pool/updates/main/o/openssl/libssl1.0.0_1.0.1t-1+deb8u12_amd64.deb && \
+   apt-dpkg-wrap dpkg -i /ssl.deb && \
+   rm /ssl.deb -f
+
```
Comment (sapkra) [inline jvb/Dockerfile:19]: What about using the docker `ADD`?

```suggestion
# https://github.com/jitsi/docker-jitsi-meet/issues/207
ADD http://security.debian.org/debian-security/pool/updates/main/o/openssl/libssl1.0.0_1.0.1t-1+deb8u12_amd64.deb /ssl.deb
RUN apt-dpkg-wrap dpkg -i /ssl.deb && rm -f /ssl.deb
   
```

File: env.example [hunk 2]
```diff
@@ -332,3 +332,18 @@ JIBRI_LOGS_DIR=/config/logs
...
+JICOFO_SHORT_ID=1
+# for testing --> SplitBridgeSelectionStrategy
+# for production --> RegionBasedBridgeSelectionStrategy
+OCTO_BRIDGE_SELECTION_STRATEGY=SplitBridgeSelectionStrategy
+
+# JVB Octo config. To enable this config set `TESTING_OCTO_PROBABILITY=1`
+JVB_NAT_HARVESTER_LOCAL_ADDRESS="0.0.0.0"
+# set the public-IP of the instance on which JVB is running
+JVB_NAT_HARVESTER_PUBLIC_ADDRESS  
+# Incase running multiple JVB's, change this value
```
Comment (sapkra) [inline env.example:346]:
```suggestion
# In case of running multiple JVB's, change this value
```

File: examples/traefik-v2/README.md [hunk 3]
```diff
@@ -1,18 +1,14 @@
...
-Note: Tested with traefik 2.2.0
+Note: Tested with traefik 2.3.1
 
 - When running behind traefik, it's a better practice to remove the port-binds for the web service.
-- The provided example uses an external network with the name "web". This is the network which moste likely was created while setting up traefik.
-- Look for comments starting with **#traefik:** to see the changes made in docker-compose.yml.
-- Traefik obtains Let's Encrypt certificates automatically.
-
-Uncomment and set DOCKER_HOST_ADDRESS in .env. I'm pretty sure, that this is mandatory for the docker-setup and should be clearer in the original README. Could be the proxying, didn't investigate further.
+- The provided example uses two external networks with the name `jitsi` for `jitsi-setup` and `proxy` for `Traefik`. Both of these network need to be created externally using below commands.
```
Comment (sapkra) [inline examples/traefik-v2/README.md:6]:
```suggestion
- The provided example uses two external networks with the name `jitsi` for `jitsi-setup` and `proxy` for `Traefik`. Both of these networks need to be created externally using the below commands.
```

File: examples/traefik-v2/README.md [hunk 4]
```diff
@@ -1,18 +1,14 @@
...
+  **`docker network create --driver=bridge jitsi`**
 
-Add or rewrite the example with docker-compose extends
+  **`docker network create --driver=bridge proxy`**
 
-````env
-DOCKER_HOST_ADDRESS=1.2.3.4
-````
+- Look for comments starting with **#traefik:** to see the changes made in docker-compose.yml.
+- Traefik obtains Let's Encrypt certificates automatically. This example uses `TLS-challenge` for obtaining certificate.
```
Comment (sapkra) [inline examples/traefik-v2/README.md:13]:
```suggestion
- Traefik obtains Let's Encrypt certificates automatically. This example uses `TLS-challenge` to obtain the certificate. 
```

**Comment by tusharsonawanes** [comment]
Hello. Thank you soo much @prayagsingh and @sapkra for this much needed repo. I have cloned it and I am facing some challenges in the same which I will mention down below.

My Infrastructure looks something like this

Machine | IP | Role
-- | -- | --
A | 172.16.4.200 | Full Docker Jitsi Meet Stack (The current repo one #804)
B | 172.16.4.231 | JVB1
C | 172.16.4.232 | JVB2
D | 172.16.4.233 | JVB3

Now I have a stack successfully running on Machine A. From what I understand, the OCTO configurations and settings have been taken care of already in the repo. Correct me if I a wrong, but what I think I need to do is, 

1. Clone the same repo on my other three machines B,C and D 
2. Change the .env argument ```JVB_NAT_HARVESTER_PUBLIC_ADDRESS="<IP of machine A>" on B,C and D
3. In docker compose, just keep jvb container configs. 
4. Start the jvb containers on B C and D.  ( And the JVB's will auto-register themselves to prosody of machine A) Am I right ? Have I missed any steps ?

**Comment by prayagsingh** [comment]
@tusharsonawanes which setup are you using docker, compose, swarm or k8s for deployment?

**Comment by tusharsonawanes** [comment]
@prayagsingh that was quick! Thank you

Plain docker on all the for machines

**Comment by tusharsonawanes** [comment]
@prayagsingh btw, JVB runs on machine 1 also. I have edited what I have written above . Machine B has jvb 2, Machine 3 has JVB 3 and Machine 4 has JVB 4.

**Comment by prayagsingh** [comment]
@tusharsonawanes  sorry I can't help you with plain docker setup. I never tried Octo with docker. I would recommend you to go for orchestration like swarm or k8s that how I deployed.

**Comment by tusharsonawanes** [comment]
@prayagsingh no issues, I will configure swarm on all the four machines. A will be the leader, and I will join the rest B, C and D to A. 

First I need to down the stack, give me 5 minutes

**Comment by tusharsonawanes** [comment]
Done. A Is the leader, with no containers running on it, and B C and D joined swarm cluster. Now how do we go ahead ?

**Comment by prayagsingh** [comment]
Please create a thread in jitsi community(install/docker) and let's move our discussion over there.

**Comment by tusharsonawanes** [comment]
Best. This will help everybody.

**Comment by tusharsonawanes** [comment]
@prayagsingh please check this - https://community.jitsi.org/t/discussion-octo-setup-in-docker-jitsi-meet-instance/83537

**Comment by prayagsingh** [comment]
@saghul @sapkra Please review it. I have added the changes suggested by @sapkra

**Comment by prayagsingh** [comment]
@saghul I'll add #785 changes to this PR and we can close 785. What do you suggest?

**Comment by prayagsingh** [comment]
@saghul  This also close #698 #699 
The `Traefik` config that I shared contains a flag `--providers.docker.exposedbydefault=false` which ensure that we have to enable `Traefik` explicitly by setting `traefik.enable=true` under `labels` for any service.

**Comment by tusharsonawanes** [comment]
@prayagsingh any help ?

**Comment by prayagsingh** [comment]
@sapkra @saghul @bbaldino I have a question regarding websocket in jvb.conf file. What if I don't want to use  websocket with jvb so setting `enable=false` is enough or Do I have add some more config if I'm disabling websocket?

I'm asking this because recently I tested my setup(websocket enable) with `ClearOne Interact 8i` conference device and some how jibri didn't work. we got websocket error but when we switch to our other setup(websocket disabled) then jibri worked. 

hence I'm thing adding below check for websocket. 
```
{{ if $ENABLE_JVB_WEBSOCKET }}
    websockets {
        enabled = true
        domain = "{{ $WS_DOMAIN }}"
        tls = true
        server-id = "{{ $WS_SERVER_ID }}"
    }
{{ end }}
```

**Comment by sapkra** [comment]
It seems like we forgot to add the config for the port in the new config:

```properties
http-servers {
  # The HTTP server which hosts services intended for 'public' use
  # (e.g. websockets for the bridge channel connection)
  public {
    port = 9090
  }
}
```

**Comment by prayagsingh** [comment]
> 
> 
> @saghul It seems like we forgot to add the config for the port in the new config:
> 
> ```ini
> http-servers {
>   # The HTTP server which hosts services intended for 'public' use
>   # (e.g. websockets for the bridge channel connection)
>   public {
>     port = 9090
>   }
> }
> ```
> 
> @prayagsingh If you can tell me if it works, you or I can create a PR for it.

Sure I'll try.

**Comment by prayagsingh** [comment]
@sapkra This is how I added
```
http-servers {    
        # The HTTP server which hosts services intended for 'public' use
        # (e.g. websockets for the bridge channel connection)
        public {
            port = 9090
        }
        private {
            host = 0.0.0.0
        }
    }
```
But it isn't working on local setup(https://localhost:8443). Removing this part fixes websocket issue with same setup. 
**Image:**
Screenshot (370)

**Comment by bbaldino** [comment]
> @sapkra @saghul @bbaldino I have a question regarding websocket in jvb.conf file. What if I don't want to use websocket with jvb so setting `enable=false` is enough or Do I have add some more config if I'm disabling websocket?
> 
> I'm asking this because recently I tested my setup(websocket enable) with `ClearOne Interact 8i` conference device and some how jibri didn't work. we got websocket error but when we switch to our other setup(websocket disabled) then jibri worked.
> 
> hence I'm thing adding below check for websocket.
> 
> ```
> {{ if $ENABLE_JVB_WEBSOCKET }}
>     websockets {
>         enabled = true
>         domain = "{{ $WS_DOMAIN }}"
>         tls = true
>         server-id = "{{ $WS_SERVER_ID }}"
>     }
> {{ end }}
> ```

`enabled=false` is enough to disable websockets, but if you disable websockets you'll need to make sure sctp is enabled or else the bridge channel won't be established.

**Comment by prayagsingh** [comment]
@sapkra  I was going through web-socket [doc](https://github.com/jitsi/jitsi-videobridge/blob/master/doc/web-sockets.md) and under `videobridge/web-socket` sectionthere is a statement  `tls = true # or false, depending on your HTTP server config`

In our case `tls=true` for websockets so do you think we have to use `tls-port` for `http-servers/public` too?

**Comment by prayagsingh** [comment]
> 
> 
> > @sapkra @saghul @bbaldino I have a question regarding websocket in jvb.conf file. What if I don't want to use websocket with jvb so setting `enable=false` is enough or Do I have add some more config if I'm disabling websocket?
> > I'm asking this because recently I tested my setup(websocket enable) with `ClearOne Interact 8i` conference device and some how jibri didn't work. we got websocket error but when we switch to our other setup(websocket disabled) then jibri worked.
> > hence I'm thing adding below check for websocket.
> > ```
> > {{ if $ENABLE_JVB_WEBSOCKET }}
> >     websockets {
> >         enabled = true
> >         domain = "{{ $WS_DOMAIN }}"
> >         tls = true
> >         server-id = "{{ $WS_SERVER_ID }}"
> >     }
> > {{ end }}
> > ```
> 
> `enabled=false` is enough to disable websockets, but if you disable websockets you'll need to make sure sctp is enabled or else the bridge channel won't be established.

Thanks @bbaldino So if I disable websockets then do I have to downgrade the JVB version too?

**Comment by bbaldino** [comment]
> Thanks @bbaldino So if I disable websockets then do I have to downgrade the JVB version too?

No, SCTP support hasn't been removed yet.

**Comment by prayagsingh** [comment]
@bbaldino Can you please share how to enable it in `Jicofo` in `sip-communicator.properties` file? 
Docker setup still use `sip-communicator.properties` file for Jicofo config while non-docker setup uses `jicofo.conf`.

**Comment by bbaldino** [comment]
> @bbaldino Can you please share how to enable it in `Jicofo` in `sip-communicator.properties` file?
> Docker setup still use `sip-communicator.properties` file for Jicofo config while non-docker setup uses `jicofo.conf`.

It looks like this config in Jicofo is only available in the new config file, so we'll have to get that added to the Jicofo docker.

**Comment by prayagsingh** [comment]
> 
> 
> > @bbaldino Can you please share how to enable it in `Jicofo` in `sip-communicator.properties` file?
> > Docker setup still use `sip-communicator.properties` file for Jicofo config while non-docker setup uses `jicofo.conf`.
> 
> It looks like this config in Jicofo is only available in the new config file, so we'll have to get that added to the Jicofo docker.

Started working on it. I'll share first-cut soon.

**Comment by prayagsingh** [comment]
@saghul Is anything missing in this PR? I think it is ready for merge.

**Comment by tusharsonawanes** [comment]
@prayagsingh still unable to join other jvb. Can you help me a little ?

**Comment by tusharsonawanes** [comment]
@prayagsingh , successfully setup octo using this PR. Just a minor issue I have. Please take a look at this - https://community.jitsi.org/t/octo-minor-issue-unable-to-see-participant-video-from-another-jvb/87482

**Comment by pandel** [comment]
@prayagsingh FYI, I needed to remove your changes regarding ssl.deb in jvb/Dockerfile to successfully build docker images after merging into current master (https://github.com/jitsi/docker-jitsi-meet/commit/5679578d808e33e0eb2ffbcff67e8b643992b177)

**Comment by tusharsonawanes** [comment]
Hi guys, any update on this ?

**Comment by prayagsingh** [comment]
> 
> 
> Hi guys, any update on this ?

Some new changes are coming in jicofo config(deprecating sip.communicator file) so based on that I'll create new PR and close this probably by weekend.

**Comment by tusharsonawanes** [comment]
@prayagsingh thats great. Thanks for the reply. Just a quick question, dockerized-octo will be fully supported in the new version right ?

**Comment by saghul** [comment]
> dockerized-octo will be fully supported in the new version right ?

I don't think we'll have time to review this in time, sorry. It will come.

**Comment by tusharsonawanes** [comment]
Just waiting for dockerized OCTO.

Thank you @saghul @prayagsingh

**Comment by prayagsingh** [comment]
> Just waiting for dockerized OCTO.
> 
> Thank you @saghul @prayagsingh 

You can build your own docker image with octo config. Use this PR to config jvb.conf and build your image.

**Comment by tusharsonawanes** [comment]
@prayagsingh  yes. 

I have used your repo and configured octo successfully. Just a small issue I had. 

https://community.jitsi.org/t/octo-minor-issue-unable-to-see-participant-video-from-another-jvb/87482

Can you please help me with this ?

**Comment by sapkra** [comment]
@prayagsingh You can upgrade this PR now to use the new jicofo config. ;)

**Comment by prayagsingh** [comment]
> 
> 
> @prayagsingh You can upgrade this PR now to use the new jicofo config. ;)

I'm planning to create a fresh new PR and close this one. I have added some more config to jvb.conf(on my local setup). Please share your feedback.

**jvb.conf** 
```
{{ $JVB_TCP_PORT := .Env.JVB_TCP_PORT | default "4443" }}
{{ $JVB_TCP_MAPPED_PORT := .Env.JVB_TCP_MAPPED_PORT | default $JVB_TCP_PORT }}
{{ $PUBLIC_URL_DOMAIN := .Env.PUBLIC_URL | default "https://localhost:8443" | trimPrefix "https://" | trimSuffix "/" -}}
{{ $WS_DOMAIN := .Env.JVB_WS_DOMAIN | default $PUBLIC_URL_DOMAIN -}}
{{ $WS_SERVER_ID := .Env.JVB_WS_SERVER_ID | default .Env.LOCAL_ADDRESS -}}
{{ $ENABLE_JVB_WEBSOCKET := .Env.ENABLE_JVB_WEBSOCKET | default "1" | toBool }}
{{ $ENABLE_OCTO := .Env.TESTING_OCTO_PROBABILITY | default "0" | toBool }}

videobridge {
    ice {
        udp {
            port = {{ .Env.JVB_PORT }}
        }
        tcp {
            enabled = {{ not .Env.JVB_TCP_HARVESTER_DISABLED }}
            port = {{ .Env.JVB_TCP_PORT }}
            {{ if not (eq $JVB_TCP_PORT $JVB_TCP_MAPPED_PORT) }}
            mapped-port = {{ $JVB_TCP_MAPPED_PORT }}
            {{ end }}
        }
    }
    {{ if .Env.JVB_CC_ENABLE | default "false" | toBool -}}
    cc {
        {{ if .Env.JVB_BWE_CHANGE_THRESHOLD }}
        # def 0.15
        bwe-change-threshold={{ .Env.JVB_BWE_CHANGE_THRESHOLD }}
        {{ end }}
        {{ if .Env.JVB_THUMBNAIL_MAX_HEIGHT }}
        # def 180
        thumbnail-max-height-px={{ .Env.JVB_THUMBNAIL_MAX_HEIGHT }}
        {{ end }}
        {{ if .Env.JVB_ONSTAGE_IDEAL_HEIGHT }}
        # def 1080
        onstage-ideal-height-px={{ .Env.JVB_ONSTAGE_IDEAL_HEIGHT }}
        {{ end }}
        {{ if .Env.JVB_ONSTAGE_PREF_HEIGHT }}
        # def 360
        onstage-preferred-height-px={{ .Env.JVB_ONSTAGE_PREF_HEIGHT }}
        {{ end }}
        {{ if .Env.JVB_ONSTAGE_PREF_FRAMERATE }}
        # 30
        onstage-preferred-framerate={{ .Env.JVB_ONSTAGE_PREF_FRAMERATE }}
        {{ end }}
        {{ if .Env.JVB_ALLOW_OVERSEND_ONSTAGE }}
        # def true
        allow-oversend-onstage={{ .Env.JVB_ALLOW_OVERSEND_ONSTAGE }}
        {{ end }}
        {{ if .Env.JVB_TRUST_BWE }}
        # def true
        trust-bwe={{ .Env.JVB_TRUST_BWE }}
        {{ end }}
        {{ if .Env.JVB_CC_PADDING_PERIOD }}
        # How often we check to send probing data. def 15ms
        padding-period={{ .Env.JVB_CC_PADDING_PERIOD }}
        {{ end }}
        {{ if .Env.JVB_FORWARD_STEAM_RECALCULATION }}
        # How often we'll force recalculations of forwarded
        # streams. def 15 seconds 
        max-time-between-calculations = {{ .Env.JVB_FORWARD_STEAM_RECALCULATION }}
        {{ end }}
        {{ if .Env.JVB_LAST_N }}
        # A JVB-wide last-n value, observed by all endpoints.  Endpoints
        # will take the minimum of their setting and this one (-1 implies
        # no last-n limit)
        jvb-last-n = {{ .Env.JVB_LAST_N }}
        {{ end }}
    }
    {{ end }}

    apis {
        xmpp-client {
            configs {
                shard {
                    HOSTNAME = "{{ .Env.XMPP_SERVER }}"
                    DOMAIN = "{{ .Env.XMPP_AUTH_DOMAIN }}"
                    USERNAME = "{{ .Env.JVB_AUTH_USER }}"
                    PASSWORD = "{{ .Env.JVB_AUTH_PASSWORD }}"
                    MUC_JIDS = "{{ .Env.JVB_BREWERY_MUC }}@{{ .Env.XMPP_INTERNAL_MUC_DOMAIN }}"
                    MUC_NICKNAME = "{{ .Env.HOSTNAME }}"
                    DISABLE_CERTIFICATE_VERIFICATION = true
                }
            }
        }

        # The COLIBRI REST API
        {{ if .Env.COLIBRI_API_ENABLE | default "false" | toBool -}}
        rest {
            enabled = {{ .Env.COLIBRI_API_ENABLE }}
        }
        {{ end }}
        {{ if .Env.JVB_API_ENABLE | default "false" | toBool -}}
        jvb-api {
            enabled = {{ .Env.JVB_API_ENABLE }}
        }
        {{ end }}
    }

    {{ if .Env.ENABLE_REST_API }}
    # Configuration of the different REST APIs.
    # Note that the COLIBRI REST API is configured under videobridge.apis.rest instead.
    rest {
        {{ if .Env.REST_DEBUG_ENABLE }}
        debug {
        enabled = {{ .Env.REST_DEBUG_ENABLE }}
        }
        {{ end }}
        {{ if .Env.REST_HEALTH_ENABLE }}
        health {
        enabled = {{ .Env.REST_HEALTH_ENABLE }}
        }
        {{ end }}
        {{ if and .Env.REST_SHUTDOWN_ENABLE .Env.COLIBRI_API_ENABLE }}
        shutdown {
        # Note that the shutdown API requires the COLIBRI API to also be enabled.
        enabled = {{ .Env.REST_SHUTDOWN_ENABLE }}
        }
        {{ end }}
    }
    {{ end }}

    {{ if .Env.JVB_LOAD_MGMT_ENABLE | default "false" | toBool -}}
    load-management {
        {{ if .Env.JVB_LOAD_MGMT_REDUCER_ENABLE }}
        # Whether or not the reducer will be enabled to take actions to mitigate load. def false
        reducer-enabled = {{ .Env.JVB_LOAD_MGMT_REDUCER_ENABLE }}
        {{ end }}

        {{ if or .Env.JVB_PKT_RATE_LOAD_THRESHOLD .Env.JVB_PKT_RATE_RCVRY_THRESHOLD }}
        load-measurements {
          packet-rate {
            {{ if .Env.JVB_PKT_RATE_LOAD_THRESHOLD }}
            # The packet rate at which we'll consider the bridge overloaded. default 50K
            load-threshold = {{ .Env.JVB_PKT_RATE_LOAD_THRESHOLD }}
            {{ end }}
            {{ if .Env.JVB_PKT_RATE_RCVRY_THRESHOLD }}
            # The packet rate at which we'll consider the bridge 'underloaded' enough
            # to start recovery. default 40K
            recovery-threshold = {{ .Env.JVB_PKT_RATE_RCVRY_THRESHOLD }}
            {{ end }}
          }
        }
        {{ end }}

        {{ if .Env.JVB_LOAD_REDUCER_ENABLE | default "false" | toBool -}}
        load-reducers {
            last-n {
                {{ if .Env.JVB_LAST_N_REDUCTION_SCALE }}
                # The factor by which we'll reduce the current last-n when trying to reduce load. def 0.75
                reduction-scale = {{ .Env.JVB_LAST_N_REDUCTION_SCALE }}
                {{ end }}

                {{ if .Env.JVB_LAST_N_RECOVER_SCALE }}
                # The factor by which we'll increase the current last-n when trying to recover. def 1.25
                recover-scale = {{ .Env.JVB_LAST_N_RECOVER_SCALE }}
                {{ end }}

                {{ if .Env.JVB_LAST_N_IMPACT_TIME }}
                # The minimum time in between runs of the last-n reducer to reduce or recover from
                # load. def 1 min
                impact-time = {{ .Env.JVB_LAST_N_IMPACT_TIME }}
                {{ end }}

                {{ if .Env.JVB_MIN_LAST_N_VAL }}
                # The lowest value we'll set for last-n. def 1
                minimum-last-n-value = {{ .Env.JVB_MIN_LAST_N_VAL }}
                {{ end }}

                {{ if .Env.JVB_MAX_ENFORCED_LAST_N_VAL }}
                # The highest last-n value we'll enforce.  Once the enforced last-n exceeds this value
                # we'll remove the limit entirely. def 40
                maximum-enforced-last-n-value = {{ .Env.JVB_MAX_ENFORCED_LAST_N_VAL }}
                {{ end }}
            }
        }
        {{ end }}
    }
    {{ end }}

    stats {
        enabled = true
    }
    {{ if $ENABLE_JVB_WEBSOCKET }}
    websockets {
        enabled = true
        domain = "{{ $WS_DOMAIN }}"
        tls = true
        server-id = "{{ $WS_SERVER_ID }}"
    }
    {{ else }}
    sctp {
        # Whether SCTP data channels are enabled.
        enabled=true
    }
    {{ end }}  

    http-servers {
        private { 
          host = 0.0.0.0
        }
        public {
            host = 0.0.0.0
            port = 9090
        }
    }
    {{ if $ENABLE_OCTO }}
    octo {
        enabled = true
        bind-address = "{{ .Env.JVB_NAT_HARVESTER_LOCAL_ADDRESS | default "0.0.0.0" }}"
        public-address = "{{ .Env.JVB_NAT_HARVESTER_PUBLIC_ADDRESS }}"
        bind-port = "{{ .Env.OCTO_BIND_PORT | default "4096" }}"
        region = "{{ .Env.OCTO_REGION | default "Region1" }}"
    }
    {{ end }}
}

ice4j {
    harvest {
        mapping {
            stun {
                {{ if .Env.JVB_STUN_SERVERS }}
                addresses = [ "{{ join "\",\"" (splitList "," .Env.JVB_STUN_SERVERS) }}" ]
                {{ end }}
            }
        }
    }
}

```

**Comment by saghul** [comment]
@prayagsingh I'd recommend starting with the smallest PR possible, that will ensure a faster review.

**Pull request closed without merging**